### PR TITLE
Nca temp

### PR DESCRIPTION
**Pull request by wdevazelhes**

### CODE REVIEW

File: sklearn/neighbors/nca.py [hunk 1]
```diff
@@ -0,0 +1,524 @@
...
+    KNeighborsClassifier(...)
+    >>> print(knn.score(nca.transform(X_test), y_test))
+    0.961904761905
+
+    Notes
+    -----
+    Neighborhood Component Analysis (NCA) is a machine learning algorithm for
+    metric learning. It learns a linear transformation in a supervised fashion
+    to improve the classification accuracy of a stochastic nearest neighbors
+    rule in the new space.
```
Comment (bellet) [inline sklearn/neighbors/nca.py]: transformed space

File: sklearn/neighbors/nca.py [hunk 2]
```diff
@@ -0,0 +1,524 @@
...
+
+    .. warning::
+
+        As NCA is optimizing a non-convex objective function, it will
+        likely end up in a local optimum. Several runs with independent random
+        init might be necessary to get a good convergence.
```
Comment (bellet) [inline sklearn/neighbors/nca.py]: Is this a standard warning used in other sklearn classes? LMNN does not have it (and is also nonconvex when solving for the linear transformation)

File: sklearn/neighbors/nca.py [hunk 3]
```diff
@@ -0,0 +1,524 @@
...
+
+    References
+    ----------
+    .. [1] J. Goldberger, G. Hinton, S. Roweis, R. Salakhutdinov.
+           "Neighbourhood Components Analysis". Advances in Neural Information
+           Processing Systems. 17, 513-520, 2005.
+           http://www.cs.nyu.edu/~roweis/papers/ncanips.pdf
+    """
```
Comment (bellet) [inline sklearn/neighbors/nca.py]: you can also add a reference to the Wikipedia page
https://en.wikipedia.org/wiki/Neighbourhood_components_analysis

File: sklearn/neighbors/nca.py [hunk 4]
```diff
@@ -0,0 +1,524 @@
...
+
+    def __init__(self, n_features_out=None, init='identity', max_iter=50,
```
Comment (bellet) [inline sklearn/neighbors/nca.py]: default `init` for LMNN is `pca`, it is probably better to have the same behavior for consistency

File: sklearn/neighbors/nca.py [hunk 5]
```diff
@@ -0,0 +1,524 @@
...
+from ..base import BaseEstimator, TransformerMixin
+from ..preprocessing import LabelEncoder
+from ..decomposition import PCA
+from ..utils.multiclass import check_classification_targets
+from ..utils.random import check_random_state
+from ..utils.validation import check_is_fitted, check_array, check_X_y
+from ..externals.six import integer_types
+
+
+class NeighborhoodComponentAnalysis(BaseEstimator, TransformerMixin):
```
Comment (bellet) [inline sklearn/neighbors/nca.py]: I think the original name in the paper (also used in the Wikipedia page) has an 's' at the end of component. probably to change the class name and other occurrences

File: sklearn/neighbors/nca.py [hunk 6]
```diff
@@ -0,0 +1,524 @@
...
+        self.random_state_ = check_random_state(self.random_state)
+
+        # Measure the total training time
+        t_train = time.time()
+
+        # Compute arrays that stay fixed during optimization:
+        # mask for fast lookup of same-class samples
+        masks = _make_masks(y_valid)
+        # pairwise differences
+        diffs = X_valid[:, np.newaxis] - X_valid[np.newaxis]
```
Comment (bellet) [inline sklearn/neighbors/nca.py]: this has dimension `(n_samples, n_samples, n_features)`, hence uses a lot of memory for large datasets. do we really gain a lot by precomputing this for not so big datasets? if not, maybe precomputing this is not necessary and you can simply work directly with `X_embedded` (which makes the gradient computation more clear I think, as you would not "propagate" the transformation matrix). if on the other hand it provides a large gain, maybe there could be an option with a simple heuristic on the dataset size/dimension to choose the default. this is probably not an urgent concern though

File: sklearn/neighbors/nca.py [hunk 7]
```diff
@@ -0,0 +1,524 @@
...
+        Data samples labels.
+
+    Returns
+    -------
+    masks: array, shape (n_samples, n_classes)
+        One-hot encoding of ``y``.
+    """
+
+    n = y.shape[0]
+    masks = np.zeros((n, y.max() + 1))
```
Comment (bellet) [inline sklearn/neighbors/nca.py]: you should make this boolean in the first place

Comment (bellet) [reply sklearn/neighbors/nca.py]: also should check whether `sklearn.preprocessing.LabelBinarizer` may be used instead? (not sure about the case of two classes). might also be better to use a sparse matrix

File: sklearn/neighbors/nca.py [hunk 8]
```diff
@@ -0,0 +1,545 @@
...
+        NotFittedError
+            If :meth:`fit` has not been called before.
+        """
+
+        check_is_fitted(self, ['transformation_'])
+        X = check_array(X)
+
+        return np.dot(X, self.transformation_.T)
+
+    def _validate_params(self, X, y):
```
Comment (bellet) [inline sklearn/neighbors/nca.py]: some questions on parameter checks:
- are we robust to the case where there is only one class? In this case there is nothing to do (the objective function is 0) so we could rule out this case when checking parameters
- are we robust to singleton classes?

File: sklearn/neighbors/nca.py [hunk 9]
```diff
@@ -0,0 +1,545 @@
...
+        ----------
+        transformation : array, shape(n_features_out, n_features)
+            The solution computed by the optimizer in this iteration.
+        """
+        if self.callback is not None:
+            self.callback(transformation, self.n_iter_)
+
+        self.n_iter_ += 1
+
+    def _loss_grad_lbfgs(self, transformation, X, y, diffs,
```
Comment (bellet) [inline sklearn/neighbors/nca.py]: one can use a `sign` argument when maximizing a function, maybe more elegant than returning - the loss and the gradient:
https://docs.scipy.org/doc/scipy/reference/tutorial/optimize.html#constrained-minimization-of-multivariate-scalar-functions-minimize

also when verbose is activated, is the objective value printed in negative or positive form? it would be best to print in positive form (as it is the original objective function of the paper).

File: sklearn/neighbors/nca.py [hunk 10]
```diff
@@ -0,0 +1,545 @@
...
+        # for every sample, compute its contribution to loss and gradient
+        for i in range(X.shape[0]):
+            diff_embedded = X_embedded[i] - X_embedded
+            sum_of_squares = np.einsum('ij,ij->i', diff_embedded,
+                                       diff_embedded)
+            sum_of_squares[i] = np.inf
+            soft = np.exp(-sum_of_squares - logsumexp(-sum_of_squares))
+            ci = masks[:, y[i]]
+            p_i_j = soft[ci]
+            not_ci = np.logical_not(ci)
```
Comment (bellet) [inline sklearn/neighbors/nca.py]: can't we just use `~ci` and avoid defining this variable?

File: sklearn/neighbors/nca.py [hunk 11]
```diff
@@ -0,0 +1,545 @@
...
+            diff_ci = diffs[i, ci, :]
+            diff_not_ci = diffs[i, not_ci, :]
+            sum_ci = diff_ci.T.dot(
```
Comment (bellet) [inline sklearn/neighbors/nca.py]: again, `sum_ci` and `sum_not_ci` are distances

File: sklearn/neighbors/nca.py [hunk 12]
```diff
@@ -0,0 +1,545 @@
...
+
+        transformation = transformation.reshape(-1, X.shape[1])
+        loss = 0
+        gradient = np.zeros(transformation.shape)
+        X_embedded = transformation.dot(X.T).T
+
+        # for every sample, compute its contribution to loss and gradient
+        for i in range(X.shape[0]):
+            diff_embedded = X_embedded[i] - X_embedded
+            sum_of_squares = np.einsum('ij,ij->i', diff_embedded,
```
Comment (bellet) [inline sklearn/neighbors/nca.py]: this variable contains distances (in embedded space), maybe it's good to have this in the variable name for clarity

File: sklearn/neighbors/nca.py [hunk 13]
```diff
@@ -0,0 +1,545 @@
...
+                                       diff_embedded)
+            sum_of_squares[i] = np.inf
+            soft = np.exp(-sum_of_squares - logsumexp(-sum_of_squares))
```
Comment (bellet) [inline sklearn/neighbors/nca.py]: again maybe find a name which makes it clear that these are exponentiated distances

File: sklearn/neighbors/nca.py [hunk 14]
```diff
@@ -0,0 +1,545 @@
...
+
+        t_funcall = time.time()
+
+        transformation = transformation.reshape(-1, X.shape[1])
+        loss = 0
+        gradient = np.zeros(transformation.shape)
+        X_embedded = transformation.dot(X.T).T
+
+        # for every sample, compute its contribution to loss and gradient
+        for i in range(X.shape[0]):
```
Comment (bellet) [inline sklearn/neighbors/nca.py]: maybe a bit more comments describing the steps below would be useful, since this is the heart of the algorithm

**Review by bellet — COMMENTED** [review]
Here are some comments!

File: sklearn/neighbors/nca.py (hunk 5 above)
Comment (wdevazelhes) [reply sklearn/neighbors/nca.py]: Done

File: sklearn/neighbors/nca.py (hunk 1 above)
Comment (wdevazelhes) [reply sklearn/neighbors/nca.py]: Done

File: sklearn/neighbors/nca.py (hunk 2 above)
Comment (wdevazelhes) [reply sklearn/neighbors/nca.py]: You are right, I just saw it for k-means as a note, and thought it could be useful, but warning is surely too much, and even a note may not be necessary, for consistency with lmnn.

File: sklearn/neighbors/nca.py (hunk 3 above)
Comment (wdevazelhes) [reply sklearn/neighbors/nca.py]: Done

File: sklearn/neighbors/nca.py (hunk 4 above)
Comment (wdevazelhes) [reply sklearn/neighbors/nca.py]: Done

File: sklearn/neighbors/nca.py (hunk 6 above)
Comment (wdevazelhes) [reply sklearn/neighbors/nca.py]: All right I agree, I will run later some benchmark to evaluate this and find the heuristic

File: sklearn/neighbors/nca.py (hunk 8 above)
Comment (wdevazelhes) [reply sklearn/neighbors/nca.py]: I added a test for only one class and singleton classes, and they both pass. For one class I checked that if init='identity', the transformed X will be the same as X, and for singleton classes I just tested that "fit" does not throw any error (took the test from lmnn, except that lmnn should throw an error in the "one non-singleton" case)

File: sklearn/neighbors/nca.py (hunk 9 above)
Comment (wdevazelhes) [reply sklearn/neighbors/nca.py]: Done

File: sklearn/neighbors/nca.py (hunk 14 above)
Comment (wdevazelhes) [reply sklearn/neighbors/nca.py]: I added some, but I will continue to do so, maybe with big O notations etc ...

File: sklearn/neighbors/nca.py (hunk 12 above)
Comment (wdevazelhes) [reply sklearn/neighbors/nca.py]: Done: I called them dist_embedded

File: sklearn/neighbors/nca.py (hunk 13 above)
Comment (wdevazelhes) [reply sklearn/neighbors/nca.py]: Done: I called them exp_dist_embedded

File: sklearn/neighbors/nca.py (hunk 10 above)
Comment (wdevazelhes) [reply sklearn/neighbors/nca.py]: Yes you are right, I changed it

File: sklearn/neighbors/nca.py (hunk 7 above)
Comment (wdevazelhes) [reply sklearn/neighbors/nca.py]: Indeed, I could not make LabelBinarizer work with two classes. However, I could use OneHotEncoder for the same purpose.
What is more, I found that if I use a sparse matrix, when filtering on class `ci` like this: `ci = masks[:, y[i]]`, and then `p_i_j = exp_dist_embedded[ci]`, `ci` should be dense (I found that boolean indexing does not work if the boolean mask is sparse), so I kept it dense from the beginning

File: sklearn/neighbors/nca.py (hunk 8 above)
Comment (bellet) [reply sklearn/neighbors/nca.py]: you could also test that when all classes are singletons, that the transformed X is also the same as X

Comment (wdevazelhes) [reply sklearn/neighbors/nca.py]: You are right. I added the test, and it passed.

File: sklearn/neighbors/tests/test_nca.py [hunk 15]
```diff
@@ -0,0 +1,310 @@
...
+    # Initialize `transformation`, `X` and `y` and `NCA`
+    random_state = check_random_state(0)
+    n_features = 10
+    num_dims = 2
+    n_samples = 100
+    n_labels = 3
+    y = random_state.randint(0, n_labels, (n_samples))
+    point = random_state.randn(num_dims, n_features)
+    X = random_state.randn(n_samples, n_features)
+    nca = NeighborhoodComponentsAnalysis(None, init=point)
```
Comment (nvauquie) [inline sklearn/neighbors/tests/test_nca.py]: A quoi sert ce premier paramètre None ?

Comment (wdevazelhes) [reply sklearn/neighbors/tests/test_nca.py]: Ah oui effectivement il ne devrait pas être là... Je crois qu'il est apparu en refactorant
